### PR TITLE
Tame gradle logging + get heapdumps on OOM

### PR DESCRIPTION
**Pull request by jeremyk-91**
See #1392 for more context. This should reduce the frequency of the OOM flakes which would sometimes occur on container 5 in external CircleCI.

This change:
* sets `atlasdb-tests-shared` and `lock-impl` tests to log at warn as opposed to debug (which is the default without a `logback.xml`).
* takes a heap dump if an OOM occurs on Circle, and stores it as a build artifact. This should be safe as there's [no overhead](https://docs.oracle.com/javase/7/docs/webnotes/tsg/TSG-VM/html/clopts.html) (see B.1.2) with running with `XX:+HeapDumpOnOutOfMemoryError`. I refrained from adding the rest of the memory debugging flags e.g. `-verbose:gc` etc due to concern about making build times longer.

Key conclusions from the ticket #1392:
* This is probably caused by Gradle internally buffering log messages and not disposing them (or not disposing them in a timely fashion). 
  - ~25% heap is retained by `org.gradle.internal.remote.internal.hub.queue.MultiEndPointQueue`
  - `select count(heap.objects('java.lang.String'), "s.toString().startsWith('18:')")` in jvisualvm returns `775604`, which is very close to our `776271` total
  - We considered the possibility of this being an [SLF4J memory leak](http://jira.qos.ch/browse/SLF4J-364), but we use Gradle 3.2.1 which uses SLF4J 1.7.21, so this is not the issue.
* Root cause is probably simply that _more tests were added_; we thus became dependent on whether Gradle is able to flush its internal buffers fast enough for it to stay within the memory limit.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1395)

### CODE REVIEW

**Comment by jeremyk-91** [comment]
Verified that this works not just for the GC overhead limit case, but also legitimate OOMs (had [a test](https://circleci.com/gh/palantir/atlasdb/5617#usage-queue/containers/5) which just allocates byte arrays of quadratic size until failure, which indeed generated the hprof we want).

Merging!

File: circle.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ machine:
     version:
       oraclejdk8
   environment:
-    _JAVA_OPTIONS: "-Xmx512M"
+    _JAVA_OPTIONS: "-Xmx512M -XX:+HeapDumpOnOutOfMemoryError"
```
Comment (schlosna) [inline circle.yml]: I'd recommend we still enable GC logging as it is low overhead and invaluable when GC & memory issues arise:

```
-verbose:gc -XX:+PrintGC -XX:+PrintGCDateStamps -XX:+PrintGCDetails -XX:-TraceClassUnloading -Xloggc:build-%t-%p.gc.log
```

File: circle.yml [hunk 2]
```diff
@@ -56,6 +56,10 @@ test:
         parallel: true
     - rsync -uav "build/reports/profile" "$CIRCLE_ARTIFACTS/":
         parallel: true
+    - mkdir -p $CIRCLE_ARTIFACTS/heapdumps:
```
Comment (schlosna) [inline circle.yml]: not sure you want to run the mkdir in parallel with the find, you probably want to make sure the `$CIRCLE_ARTIFACTS/heapdumps` is created before the `find` runs

Comment (SerialVelocity) [inline circle.yml]: That's not what `parallel` does. `parallel` just runs it on all containers. `background` does what you're thinking of (see higher up in the `circle.yml`)

File: circle.yml (hunk 1 above)
Comment (jeremyk-91) [inline circle.yml]: (Done in #1398)

**Pull request merged**